File: ALODWebUtility.CS/Printing/FormFieldParsers/AdobeLiveCycleHierarchicalParser.cs
using ALOD.Core.Interfaces;
using WebSupergoo.ABCpdf8;
using WebSupergoo.ABCpdf8.Objects;

namespace ALODWebUtility.Printing.FormFieldParsers
{
    public class AdobeLiveCycleHierarchicalParser : IPDFFormFieldParser
    {
        public Field ParseField(Doc document, string desiredField)
        {
            string[] fieldNames = document.Form.GetFieldNames();
            Field currentField = null;

            foreach (string fieldName in fieldNames)
            {
                currentField = document.Form[fieldName];

                if (ExtractFieldName(currentField).Equals(desiredField))
                {
                    return currentField;
                }
            }

            return null;
        }

        protected string ExtractFieldName(Field fieldObj)
        {
            // Adobe LiveCycle field name format: form#[#].Page#[#].fieldName[#]
            // Field.PartialName property grabs the fieldName[#] portion for the field name
            return fieldObj.PartialName.Substring(0, fieldObj.PartialName.Length - 3);
        }
    }
}

File: ALODWebUtility/Printing/FormFieldParsers/AdobeLiveCycleHierarchicalParser.vb
Imports ALOD.Core.Interfaces
Imports WebSupergoo.ABCpdf8
Imports WebSupergoo.ABCpdf8.Objects

Namespace Printing.FormFieldParsers

    Public Class AdobeLiveCycleHierarchicalParser
        Implements IPDFFormFieldParser

        Public Function ParseField(ByVal document As Doc, ByVal desiredField As String) As Field Implements IPDFFormFieldParser.ParseField
            Dim fieldNames As String() = document.Form.GetFieldNames()
            Dim currentField As Field = Nothing

            For Each fieldName As String In fieldNames
                currentField = document.Form(fieldName)

                If (ExtractFieldName(currentField).Equals(desiredField)) Then
                    Return currentField
                End If
            Next

            Return Nothing
        End Function

        Protected Function ExtractFieldName(ByVal fieldObj As Field) As String
            ' Adobe LiveCycle field name format: form#[#].Page#[#].fieldName[#]
            ' Field.PartialName property grabs the fieldName[#] portion for the field name
            Return fieldObj.PartialName.Substring(0, fieldObj.PartialName.Length - 3)
        End Function

    End Class

End Namespace